Copy byron core to create shelley core

TimSheard opened the pull request:
This PR is the first step in a series of four planned PRs to replace Byron.Spec.Ledger.Core, with a new version Shelley.Spec.Ledger.Core; The new version will be a stripped down version, implementing only the functionality needed for Shelley, and with more attention paid to efficient implementations. In all but the last PR, the API of the Relation class, will remain the same.
In this first PR the Shelley version is an exact copy of the Byron version, except its module name has been changed to record that appears in a different place in the source tree.

Future PRs will address the following

step 1) Restrict the export list, and delete the unneeded code.
step 2) Rewrite the needed operations so that they pay attention to efficiency.
step 3) Rework some of the calls to the API, to take advantage of more efficient versions
step 3) Rework the API. Change some inputs from (Foldable f => f a) to (Set a)
        A foldable input forces linear behavior.

Commit:
rename byron core to shelley core

JaredCorduan reviewed (approved):
looks good to me!

Commit:
ormolu

erikd commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Validation.hs:
```diff
@@ -16,7 +16,7 @@ module Shelley.Spec.Ledger.API.Validation
 where
 
-import Byron.Spec.Ledger.Core (Relation (..))
+import Shelley.Spec.Ledger.Core (Relation (..))
```
Wait, what? Why?

These names mean something to consumers of this library, eg me. 

Was this named incorrectly? If so fine, but the commit message should state that.

erikd replied:
Furthermore at least half of PR comment above belongs in the git commit message so that the reasons for the change are recorded with the commit itself rather than with the merge commit. Everything other than future plans should be in the commit message.

JaredCorduan replied:
> Was this named incorrectly? If so fine, but the commit message should state that.

Yes, the plan is to divorce this module from Byron so that we can implement the functions the way we want to (ie efficiently) without having to change Byron. I had thought this was clear from the PR message. I like the idea of putting these notes in the commit message!

JaredCorduan replied:
To be clear, the Byron module `Byron.Spec.Ledger.Core` hasn't moved or changed. It has been copied and renamed for Shelley as a baby step so that we can dramatically strip it down and remove things that we don't want. (such as the `Ord` constraint on `Relation` so that we don't require the _values_ of a map to have `Ord`, etc).

erikd commented:
> It has been copied and renamed for Shelley

I would really appreciate it if that was made clear in the commit message and the PR message.

TimSheard commented:
> > It has been copied and renamed for Shelley
> 
> I would really appreciate it if that was made clear in the commit message and the PR message.

Erik, We changed the commit message to make very clear what is going, and that the old Byron Core module still remains unchanged.
